[클린코드 2기 표후동] 계산기 미션

ndopy opened the pull request:
안녕하세요. [Clean Code with JavaScript 2기] 진행중인 표후동입니다.
리뷰 요청드립니다.

### 진행하면서 느낀 점 & 질문
1. 코드를 작성하는 시간보다 어떻게 구현해야 할지 생각하는 시간이 훨씬 더 길어서 구현이 오래걸리는 것 같습니다.
2. 미션을 진행할 때 어떤 식으로 접근하면 좋을지?
3. 기능들을 함수(function)로 작성하였는데 이런 방식이 괜찮은지요? (모던하지 않다거나...)

ndopy committed:
계산기 기능 구현

Tanney-102 commented on src/js/calculator/calculateOperation.js:
```diff
@@ -0,0 +1,23 @@
+import { isValidOperation } from "./validateOperation.js";
+
+function calculateOperation() {
+  const usedOperator = isValidOperation();
```
isValidOpration은 boolean을 반환할 것 같은 이름인데 다른 걸 반환하나보네요~

ndopy replied:
코드를 다시 보니 이름도 어색하고 실행 내용도 중복되거나 맞지 않는 코드들이 있네요 ㅠㅠ 수정해보겠습니다.

Tanney-102 commented on src/js/calculator/calculateOperation.js:
```diff
@@ -0,0 +1,23 @@
+  } else if (usedOperator === "X") {
+    OPERATION.innerText = operand1 * operand2;
+  } else {
+    OPERATION.innerText = operand1 / operand2;
```
소숫점 이하는 버림한다는 요구사항이 있지 않았나요?

ndopy replied:
꼼꼼하지 못했네요 ㅠ 요구사항에 맞도록 수정해보겠습니다.

Tanney-102 commented on src/js/calculator/calculateOperation.js:
```diff
@@ -0,0 +1,23 @@
+  }
+}
+
+export { calculateOperation };
```
named export를 하는 이유가 있을까요?

ndopy replied:
export, import 관련 내용을 공부할 때 named, default 중 하나만 쓰는 경우가 많다고 본 적이 있습니다.
그 이후로 통일성 있게 named export 로 작성하는 게 습관이 되어 있어서 그런 것 같습니다.

Tanney-102 replied:
파일이름도 calculateOperation이고 export 하는 내용도 calculateOpration인 것 같아 질문했습니다~
명확하게 하나의 요소를 메인으로 export 할 목적이라면 default export로 의도를 밝히는 것도 좋을 것 같습니다.

Tanney-102 commented on src/js/calculator/validateOperation.js:
```diff
@@ -0,0 +1,44 @@
+    digitLength = currentOperation.length + 1;
+  }
+
+  if (digitLength > 3) {
+    alert("숫자는 3자리까지만 입력할 수 있습니다.");
+    return false;
+  }
+  return true;
```
함수 이름은 isOver3Digits인데 자릿수가 3보다 클때 false를 반환하네요

ndopy replied:
그렇네요.. 함수 이름대로라면 3자리보다 클 때 true가 되어야 할 것 같은데 실행 내용은 그 반대로 되어 있네요.
실행 내용에 맞는 네이밍으로 고쳐보겠습니다.

Tanney-102 commented on src/js/index.js:
```diff
@@ -0,0 +1,14 @@
+import { calculateOperation } from "./calculator/calculateOperation.js";
+
+const numberBtn = document.querySelector(".digits");
+const operatorBtn = document.querySelector(".operations");
```
equalSign도 .operations 안에 있지않나요?
equalSign을 클릭하면 inputOperator도 같이 호출 될 것 같습니다~

ndopy replied:
`inputOperator` 는 연산자를 클릭했을 때 입력하는 기능만 수행하는 용도로 작성했습니다.
`=` 연산자는 식을 계산하는 기능을 담당해야 하기 때문에 `calculateOperation.js` 에서 따로 작성을 했습니다.

그래서 `=` 연산자의 기능을 inputOperator 에서 제외하기 위해
`inputOperator` 에서 if문으로 조건을 걸어 `return false` 로 작성했는데
다시 보니까 굳이 `return false` 가 아닌 `calculateOperation` 을 실행하면 될 것 같습니다.

이렇게 `inputOperator` 안에서 `+, -, X, /` 와 `=` 를 분리해도 괜찮을까요?

Tanney-102 replied:
html 문서상에서 '='도 operation으로 함께 묶여있기때문에 inputOperator 안에서 처리를 해주는 것도 꽤 합리적이라는 생각이 들긴해요~
inputOperator => if '=' => calculate 요런 흐름도 괜찮아보입니다!
이렇게하면 불필요한 함수 호출을 하나 줄일 수 있겠네요~

Tanney-102 commented on src/js/calculator/inputOperator.js:
```diff
@@ -0,0 +1,25 @@
+  const currentOperation = OPERATION.innerText;
+
+  if (clickedOperator === "=") {
+    return false;
```
false를 반환하는 이유가 있을가요?

ndopy replied:
`=` 와 다른 연산자들을 분리해서 급하게 실행 코드를 작성하려고 하다가 저렇게 작성했네요 ㅠㅠ
false 를 반환하지 않고 `calculateOperation` 을 실행하는 내용으로 수정해보려고 합니다.

이렇게 `inputOperator` 안에서 `+, -, X, /` 와 `=` 를 분리해도 괜찮을까요?

Tanney-102 replied:
아아 return을 하시는 이유는 알겠는데 굳이 'false'를 반환하시는 목적이 있는지 궁금했습니다!

Tanney-102 commented on src/js/calculator/calculateOperation.js:
```diff
@@ -0,0 +1,23 @@
+  const operand1 = Number(operations[0]);
+  const operand2 = Number(operations[1]);
+
+  if (usedOperator === "+") {
+    OPERATION.innerText = operand1 + operand2;
+  } else if (usedOperator === "-") {
+    OPERATION.innerText = operand1 - operand2;
+  } else if (usedOperator === "X") {
+    OPERATION.innerText = operand1 * operand2;
+  } else {
+    OPERATION.innerText = operand1 / operand2;
```
```suggestion
  const operationExecutor = {
   "+": (operand1, operand2) => operand1 + operand2,
   "-": (operand1, operand2) => operand1 - operand2
   "X": (operand1, operand2) => operand1 * operand2
   "/": (operand1, operand2) => operand1 / operand2 
  }
  OPERATION.innerText = operationExecutor[usedOperator]
```

if else 문을 통해 분기도 좋지만 이렇게 테이블을 만들어 단순하게 분기하는 방법도 많이 사용하니 참고해주세요~

ndopy replied:
감사합니다. 이렇게 작성하면 indent depth 도 줄어들고 가독성도 훨씬 좋네요.
앞으로도 분기 코드 작성할 때 꼭 이 방법으로도 구현해보겠습니다.

Tanney-102 reviewed (changes requested):
후동님 안녕하세요!
계산기 미션 리뷰어 오태은입니다!
간단한 미션에 어울리는 간결한 구현으로 미션 진행해주셨네요!!
고생많으셨습니다! ㅎㅎ

몇가지 코멘트 드리긴했는데, 몇몇 네이밍을 고민해보시면 좋을 것 같아요.
특히 함수 이름에 경우 역할, 반환값 등을 고려해 직관적으로 잘 지어주셨는데 몇가지 어색한 부분들이 보이네요~

질문에 대한 답변드립니다.

1. 미션을 진행할 때 어떤 식으로 접근하면 좋을지?
아무래도 vanila JS로 진행하는 미션이라 어떻게 구조를 잡고 어떻게 구현을 해야할지 막막하시죠ㅠㅠ 그 맘 이해합니다. 실제 코드를 작성하기 전 고민하는 시간이 많으셨던 것 같아요~ 아마 어떤 구조와 방식이 좋고 적절한지 따져보고 망설이는 시간이지 않았을까 싶은데, 오히려 단순하게 우선 해보는 게 도움이 될 수 있습니다. 웹 프론트엔드도 결국 입력을 처리해 출력하는 '프로그램'이기에  사용자의 입력(클릭, 폼 입력 등)에 어떻게 대응할 것인지(어떤 데이터를 변경할지, 무엇을 보여줄지)에 작성해야할 코드가 있습니다. 우선 요구사항에 맞게 이러한 코드들을 한 줄, 두 줄 작성해보시는 게 중요합니다. 그 후에 어떤 구조가 좋을지, 어떤 패턴을 적용해 볼 수 있을지 고민해도 늦지않습니다. 오히려 빈 화면보다 미흡하게나마 작성한 코드가 고민하시는데 훨씬 도움이 될겁니다. 일단 요구사항에 대한 코드를 작성하는데 집중하기위해 TDD를 시도해보셔도 좋겠네요!
그래도 여전히 감이 잡히시지 않는다면 다른 분들의 코드를 보는 것도 도움이됩니다. 다른 사람들은 어떻게 구현하는지, 그렇게 했을 때 리뷰어로부터 어떤 피드백이 오는지 등을 보는 것만으로 크게 도움되실겁니다!

2. 기능들을 함수로 작성했는데 이런 방식이 괜찮은지?
우선 각 역할에 맞게 모듈을 분리하신 노력이 보여 좋았습니다. 다만 관련된 로직들은 하나의 파일 혹은 하나의 클래스로 관리해도 좋을 것 같습니다. 지금은 단순히 함수단위로 과하게 모듈화되어있는 것 같아요~
크게 어떤 일을 하는지에 따라 모듈을 나눠봐도 좋을 것 같습니다.
(예를 들어 데이터를 받아 화면에 보여주는 역할, 사용자의 입력에 따라 데이터를 가공하는 역할 등..)

테스트 코드 작성 완료하시면 push 후 다시 한 번 리뷰 요청 눌러주세요!

ndopy committed:
fix: 나눗셈 연산시 소수점을 버리지 않는 오류 수정

ndopy committed:
chore: cypress 설치

ndopy committed:
test: 계산기 기능 테스트 코드 작성

ndopy committed:
refactor: 숫자가 3자리 이하인지 검증하는 함수명 변경

ndopy committed:
refactor: 계산식 유효성 검증 함수 수정
- 함수의 기능에 맞도록 반환값 수정
- 해당 함수를 사용하는 calculationOperation 함수 내용 수정

ndopy committed:
refactor: 등호 연산자 클릭 시 함수 동작 내용 수정
- inputOperator 에서 '='의  calculateOperation 실행.
- index.html 의 = 에 지정한 'id=equalSign'은 제거.
- id=equalSign 에 별도 지정한 이벤트는 더 이상 사용하지 않으므로  제거.

ndopy committed:
refactor: if-else를 객체 리터럴 방식으로 수정.

Tanney-102 commented on cypress/integration/calculator.spec.js:
```diff
@@ -0,0 +1,68 @@
+  });
+
+  it("1. 2개의 숫자에 대해 덧셈이 가능하다.", () => {
+    cy.get(".digit").contains("1").click();
+    cy.get(".operation").contains("+").click();
+    cy.get(".digit").contains("2").click();
+    cy.get(".operation").contains("=").click();
```
반복되는 로직은 cypress commands를 통해 함수처럼 만들어 사용하실 수 있습니다!
다음 미션에 참고해보셔요!
https://docs.cypress.io/api/cypress-api/custom-commands

ndopy replied:
감사합니다. 도움이 많이 될 것 같아요!

Tanney-102 reviewed (approved):
안녕하세요 후동님!
리뷰 드린 내용을 다 반영해주셨네요!
이 미션은 머지하도록 하겠습니다 ㅎㅎ
고생많으셨어요!
